Client implementation (#1)
ostis-ai/py-sc-client: README.md

# **py-sc-client** #

> Python implementation of client for [communication with sc-server](http://ostis-dev.github.io/sc-machine/http/websocket/). This module is tested on Python 3.8+.

Reviewer: ```suggestion
The Python implementation of the client for [communication with sc-server](http://ostis-dev.github.io/sc-machine/http/websocket/). This module is tested on Python 3.8+.
```

Reviewer 2: It's probably not a good idea to provide documentation links from another organisation. That said, we didn't host ours on GitHub Pages yet

Author: > It's probably not a good idea to provide documentation links from another organisation. That said, we didn't host ours on GitHub Pages yet

I agreed, will be changed to README with protocol description of ostis-ai organisation

> Require *websocket-client* library.

Reviewer: ```suggestion
Require the *websocket-client* library.
```

```sh
pip3 install websocket-client
```

## connect

> Connect to server by url. Run client in another thread. Can be stopped only by interrupt

Reviewer: ```suggestion
Connect to the server by the url. Run the client in another thread. Can be stopped only by interrupting
```

> *Parametrs*: url of sc-server.

Reviewer: ```suggestion
*Parametrs*: url of the sc-server.
```

```py
from sc_client import client

client.connect("ws://localhost:8090/ws_json")
```

## check_elements

> Check type of given elements.

Reviewer: ```suggestion
Check the type of given elements.
```

*Parametrs*: list of ScAddr.
*Returns*: ScType of elements.

```py
    elem_types = client.check_elements([node_addr, edge_addr, link_addr])
    elem_types[0].is_node() #True
    elem_types[1].is_edge() #True
    elem_types[2].is_link() #True
```

## create_elements

> Create given construction in KB memory.

Reviewer: ```suggestion
Create a given construction in the KB memory.
```

*Parametrs*: ScConstruction.
*Returns*: ScAddr list of construction elements.

```py
    const = ScConstruction()
    const.create_node(sc_types.NODE_CONST, 'node')
    link_content = ScLinkContent("Hello!", ScLinkContentType.STRING.value)
    const.create_link(sc_types.LINK, link_content, 'link')
    const.create_edge(sc_types.EDGE_ACCESS_CONST_POS_PERM, 'node', 'link')
    addr_list = client.create_elements(const)
    addr_list # [node_addr, link_addr, edge_addr]
```

## delete_elements

> Delete elements from KB memory.

Reviewer: ```suggestion
Delete elements from the KB memory.
```

*Parametrs*: list of ScAddr.
*Returns*: True if elements were deleted.

```py
status = client.delete_elements([node_addr, edge_addr, link_addr])
status #True
```

## set_link_contents

> Set new content to given link.

Reviewer: ```suggestion
Set new contents to the given link.
```

*Parametrs*: list of ScLinkContent.

> *Returns*: True if set link content.

Reviewer: ```suggestion
*Returns*: True if the link contents is set.
```

```py
link_content = ScLinkContent("Hello world!", ScLinkContentType.STRING.value)
link_content.addr = link_addr
status = client.set_link_contents([link_content])
status # True
```

## get_link_content

> Get content of given link.

Reviewer: ```suggestion
Get the content of the given link.
```

> *Parametrs*: ScAddr of link.

Reviewer: ```suggestion
*Parametrs*: ScAddr of the link.
```

*Returns*: ScLinkContent object. 

```py
link_content = client.get_link_content(addr_list[0])
link_content.data # link content
link_content.addr # link addr
```

## get_link_by_content

> Get link from KB by content.

Reviewer: ```suggestion
Get a link from the KB by contents.
```

> *Parametrs*: List of ScLinkContent or list of string as link content, or .

Reviewer: ```suggestion
*Parametrs*: A list of ScLinkContent or a list of strings as the link contents, or .
```

> *Returns*: List of lists of ScAddrs. 

Reviewer: ```suggestion
*Returns*: A list of lists of ScAddrs. 
```

```py

> content = client.get_link_by_content(["testing search by link content as string", "no content"])

Reviewer: ```suggestion
content = client.get_link_by_content(["testing search by the link content as a string", "no content"])
```

> content[0] # get list of link ScAddrs with given content

Reviewer: ```suggestion
content[0] # get a list of ScAddrs links with given content
```

> content[1] # get empty list if no link with given content in memory

Reviewer: ```suggestion
content[1] # get an empty list if there is no link with given content in the memory
```

```

## resolve_keynodes

> Get keynode from KB memory by idtf and ScType. If it doesn't exist, then create new one. If ScType is None, tries to find element by system identifier.

Reviewer: ```suggestion
Get a keynode from the KB memory by idtf and ScType. If it doesn't exist, then create a new one. If ScType is None, try to find an element by the system identifier.
```

*Parametrs*: ScIdtfResolveParams.

> *Returns*: ScAddr of keynode. 

Reviewer: ```suggestion
*Returns*: ScAddr of the keynode. 
```

```py
params = ScIdtfResolveParams(idtf='my_new_keynode_that_not_exist', type=sc_types.NODE_CONST)
addr = client.resolve_keynodes([params])
addr # new keynode addr

params = ScIdtfResolveParams(idtf='technology_OSTIS', type=sc_types.NODE_CONST)
addr = client.resolve_keynodes([params])
addr # keynode addr

params = ScIdtfResolveParams(idtf='my_another_keynode_that_not_exist', type=None)
addr = client.resolve_keynodes([params])
addr # None

params = ScIdtfResolveParams(idtf='nrel_format', type=None)
addr = client.resolve_keynodes([params])
addr # keynode addr
```

## template_search

> Search in KB memory by template. 

Reviewer: ```suggestion
Search in the KB memory by template. 
```

> *Parametrs*: ScTemplate or scs-temlate as string.

Reviewer: ```suggestion
*Parametrs*: An ScTemplate or an scs-template as a string.
```

*Returns*: ScTemplateResult list. 

```py
templ = ScTemplate()
templ.triple([class_node_addr, '_class_node'], sc_types.EDGE_ACCESS_VAR_POS_PERM, node_addr) # faf
templ.triple('_class_node', sc_types.EDGE_ACCESS_VAR_POS_TEMP, link_addr) # faf
templ.triple('_class_node', edge_addr, sc_types.NODE_VAR) # ffa
templ.triple(node_addr, edge_addr, sc_types.NODE_VAR, is_required=False) # ffa, can be empty
search_results = client.template_search(templ)
search_result = search_results[0]

> search_result.size() # count of elements in result construction

Reviewer: ```suggestion
search_result.size() # count of elements in the resulting construction
```

> search_result.get(0).value # get element from result by index

Reviewer: ```suggestion
search_result.get(0).value # get an element from the result by index
```

> search_result.get('_class_node').value # get element from result by alias

Reviewer: ```suggestion
search_result.get('_class_node').value # get an element from the result by alias
```

def for_each_tripple_func(src: ScAddr, edge: ScAddr, trg: ScAddr):
    ...

> search_result.for_each_triple(for_each_tripple_func) # call function for each triple in result

Reviewer: ```suggestion
search_result.for_each_triple(for_each_tripple_func) # call a function for each triple in the result
```

```

## template_generate

> Generate construction in KB memory by template.

Reviewer: ```suggestion
Generate a construction in the KB memory by template.
```

> *Parametrs*: ScTemplate or scs-temlate as string & ScTemplateGenParams.

Reviewer: ```suggestion
*Parametrs*: An ScTemplate or an scs-template as a string & ScTemplateGenParams.
```

> *Returns*: ScTemplateResult object. 

Reviewer: ```suggestion
*Returns*: An ScTemplateResult object. 
```

```py
templ = ScTemplate()
templ.triple_with_relation( # faaaf
    [main_node, '_main_node'],
    sc_types.EDGE_D_COMMON_VAR,
    [sc_types.LINK, '_link'],
    sc_types.EDGE_ACCESS_VAR_POS_PERM,
    relation_node,
)
templ.triple( # faa
    '_main_node',
    sc_types.EDGE_ACCESS_VAR_POS_TEMP,
    [sc_types.NODE_VAR, '_var_node']
)
gen_params = {'_link': link_node, '_var_node': var_node}
gen_result = client.template_generate(templ, gen_params)
```

## events_create

> Create an ScEvent in KB memory.

Reviewer: ```suggestion
Create an ScEvent in the KB memory.
```

*Parametrs*: list of ScEventParams.

> *Returns*: list of ScEvent. 

Reviewer: ```suggestion
*Returns*: list of ScEvents. 
```

```py
def event_callback(src: ScAddr, edge: ScAddr, trg: ScAddr):
    ...

event_type = common.ScEventType.ADD_OUTGOING_EDGE
event_params = ScEventParams(bounded_elem_addr, event_type, event_callback)
sc_event = client.events_create([event_params])
```

## is_event_valid

> Check is ScEvent active or not.

Reviewer: ```suggestion
Check is an ScEvent active or not.
```

*Parametrs*: ScEvent.

> *Returns*: True if event valid. 

Reviewer: ```suggestion
*Returns*: True if an event is valid. 
```

```py
status = client.is_event_valid(sc_event)
status # True
```

## events_destroy

> Destroy ScEvent in KB memory.

Reviewer: ```suggestion
Destroy an ScEvent in the KB memory.
```

> *Parametrs*: list of ScEvent.

Reviewer: ```suggestion
*Parametrs*: list of ScEvents.
```

> *Returns*: True if event was destroyed. 

Reviewer: ```suggestion
*Returns*: True if the event was destroyed. 
```

```py
status = client.events_destroy([sc_event])
status # true
```

# Classes

> Python library contains python implementation of useful classes and functions to work with sc-memory.

Reviewer: ```suggestion
The Python library contains python implementation of useful classes and functions to work with the sc-memory.
```

There are a list of classes:

 - ScKeynodes
 - ScAgent
 - ScModule

## ScKeynodes

> Singleton dictinary object which provide possibility to cache identifiers and addrs of keynodes stored in KB.

Reviewer: ```suggestion
A singleton dictionary object which provides the ability to cache identifiers and addrs of keynodes stored in the KB.
```

Create an instanse of ScKeynodes class to get access to cache:

```py
keynodes = ScKeynodes()
```

Get provided identifier:
```py

> keynodes["identifier_of_keynode"] # returns ScAddr of given identifier

Reviewer: ```suggestion
keynodes["identifier_of_keynode"] # returns an ScAddr of the given identifier
```

> keynodes["not_stored_in_kb"] # returns invalid ScAddr if identifier does not exist in memory

Reviewer: ```suggestion
keynodes["not_stored_in_kb"] # returns an invalid ScAddr if an identifier does not exist in the memory
```

```

Use _resolve_identifiers()_ to upload identifiers from _Enum_ classes:

```py
class CommonIdentifiers(Enum):
    RREL_ONE = "rrel_1"
    RREL_TWO = "rrel_2"

class QuestionStatus(Enum):
    QUESTION_INITIATED = "question_initiated"
    QUESTION_FINISHED = "question_finished"
    QUESTION_FINISHED_SUCCESSFULLY = "question_finished_successfully"
    QUESTION_FINISHED_UNSUCCESSFULLY = "question_finished_unsuccessfully"
    
keynodes.resolve_identifiers([QuestionStatus, CommonIdentifiers])
```

## ScAgent

> Class for handling single ScEvent. Define your agents like this:

Reviewer: ```suggestion
A class for handling a single ScEvent. Define your agents like this:
```

```py
class MyAgent(ScAgent):
    action = "Identifier_of_action_class"
    
    def register(self) -> ScEvent:
        # override method, must return ScEvent instance
        params = [
            MyAgent.keynodes["action_initiated"],
            common.ScEventType.ADD_OUTGOING_EDGE,
            MyAgent.run_impl
        ]
        event_params = ScEventParams(*params)
        sc_event = client.events_create([event_params])
        return sc_event[0]
    
    @staticmethod
    def run_impl(action_class: ScAddr, edge: ScAddr, action_node: ScAddr) -> None:
        # override method, must have 3 args and be static
        ...

```

## ScModule

> Class for handling multiple ScAgent. Define your modules like this:

Reviewer: ```suggestion
A class for handling a multiple ScAgent. Define your modules like this:
```

# **py-sc-client** #

The python implementation of the client for [communication with sc-server](http://ostis-dev.github.io/sc-machine/http/websocket/). This module is tested on Python 3.8+.

Require the *websocket-client* library.
```sh
pip3 install websocket-client
```

## connect
Connect to the server by the URL. Run the client in another thread.

> *Parameters*: URL of the sc-server as a string.

Reviewer: ```suggestion
*Parameters*: A URL of the sc-server as a string.
```

```py
from sc_client import client

client.disconnect()
```

## is_connected
Check the state of the connection with the sc-server.

*Returns*: The boolean value (true if a connection is established).

```py
from sc_client import client

client.is_connected() # True
```

## check_elements
Check the type of given elements.

*Parameters*: ScAddr class objects.
*Returns*: A list of ScType class objects.

```py
elem_types = client.check_elements(node_addr, edge_addr, link_addr)
elem_types[0].is_node() #True
elem_types[1].is_edge() #True
elem_types[2].is_link() #True
```

## create_elements
Create a given construction in the KB memory.

*Parameters*: An ScConstruction class object.
*Returns*: A list of ScAddr class objects.

```py
const = ScConstruction()
const.create_node(sc_types.NODE_CONST, 'node')
link_content = ScLinkContent("Hello!", ScLinkContentType.STRING.value)
const.create_link(sc_types.LINK, link_content, 'link')
const.create_edge(sc_types.EDGE_ACCESS_CONST_POS_PERM, 'node', 'link')
addr_list = client.create_elements(const)
addr_list # [node_addr, link_addr, edge_addr]
```

## delete_elements
Delete elements from the KB memory.

*Parameters*: ScAddr class objects.

> *Returns*: A boolean value (true if elements were deleted).

Reviewer: ```suggestion
*Returns*: The boolean value (true if elements were deleted).
```

```py
status = client.delete_elements(node_addr, edge_addr, link_addr)
status #True
```

## set_link_contents

> Set new contents to corresponding links.

Reviewer: ```suggestion
Set the new content to corresponding links.
```

*Parameters*: ScLinkContent class objects.

> *Returns*: A boolean value (true if the link content is set).

Reviewer: ```suggestion
*Returns*: The boolean value (true if the link content is set).
```

```py
link_content = ScLinkContent("Hello world!", ScLinkContentType.STRING.value)
link_content.addr = link_addr
status = client.set_link_contents(link_content)
status # True
```

## get_link_content
Get the content of the given link.

*Parameters*: An ScAddr class object.
*Returns*: An ScLinkContent class object. 

```py
link_content = client.get_link_content(link_addr)
link_content.data # link content
link_content.addr # link addr
```

## get_links_by_content
Get links from the KB with content.

*Parameters*: ScLinkContent class objects, strings.
*Returns*: A list of lists of ScAddr class objects.

```py
content = client.get_links_by_content("testing search by the link content as a string", "no content")

> content[0]  # get a list of ScAddrs links with given content

Reviewer: ```suggestion
content[0]  # get a list of ScAddrs links with the given content
```

> content[1]  # get an empty list if there is no link with given content in the memory

Reviewer: ```suggestion
content[1]  # get an empty list if there is no link with the given content in the memory
```

```

## resolve_keynodes
Get a keynode from the KB memory by identifier and type. If it doesn't exist, then create a new one. 
If ScType is None, try to find an element by the system identifier.

*Parameters*: An ScIdtfResolveParams class object.
*Returns*: An ScAddr class object. 

```py
params = ScIdtfResolveParams(idtf='my_new_keynode_that_not_exist', type=sc_types.NODE_CONST)
addr = client.resolve_keynodes([params])
addr # new keynode addr

params = ScIdtfResolveParams(idtf='technology_OSTIS', type=sc_types.NODE_CONST)
addr = client.resolve_keynodes([params])
addr # keynode addr

params = ScIdtfResolveParams(idtf='my_another_keynode_that_not_exist', type=None)
addr = client.resolve_keynodes([params])
addr # None

params = ScIdtfResolveParams(idtf='nrel_format', type=None)
addr = client.resolve_keynodes([params])
addr # keynode addr
```

## template_search

> Search in the KB memory by a template. 

Reviewer: ```suggestion
Search in the KB memory by template. 
```

*Parameters*: An ScTemplate class object or an scs-template as a string.
*Returns*: A list of ScTemplateResult class objects. 

```py
templ = ScTemplate()
templ.triple([class_node_addr, '_class_node'], sc_types.EDGE_ACCESS_VAR_POS_PERM, node_addr) # faf
templ.triple('_class_node', sc_types.EDGE_ACCESS_VAR_POS_TEMP, link_addr) # faf
templ.triple('_class_node', edge_addr, sc_types.NODE_VAR) # ffa
templ.triple(node_addr, edge_addr, sc_types.NODE_VAR, is_required=False) # ffa, can be empty
search_results = client.template_search(templ)
search_result = search_results[0]

search_result.size() # count of elements in the resulting construction
search_result.get(0).value # get an element from the result by index
search_result.get('_class_node').value # get an element from the result by alias

def for_each_tripple_func(src: ScAddr, edge: ScAddr, trg: ScAddr):
    ...
search_result.for_each_triple(for_each_tripple_func) # call a function for each triple in the result
```

## template_generate

> Generate a construction in the KB memory by a template.

Reviewer: ```suggestion
Generate a construction in the KB memory by template.
```

*Parameters*: An ScTemplate class object or an scs-template as a string, an ScTemplateGenParams class object.
*Returns*: An ScTemplateResult class object. 

```py
templ = ScTemplate()
templ.triple_with_relation( # faaaf
    [main_node, '_main_node'],
    sc_types.EDGE_D_COMMON_VAR,
    [sc_types.LINK, '_link'],
    sc_types.EDGE_ACCESS_VAR_POS_PERM,
    relation_node,
)
templ.triple( # faa
    '_main_node',
    sc_types.EDGE_ACCESS_VAR_POS_TEMP,
    [sc_types.NODE_VAR, '_var_node']
)
gen_params = {'_link': link_node, '_var_node': var_node}
gen_result = client.template_generate(templ, gen_params)
```

## events_create
Create an event in the KB memory.

*Parameters*: ScEventParams class objects.
*Returns*: A list of ScEvent class objects. 

```py
def event_callback(src: ScAddr, edge: ScAddr, trg: ScAddr):
    ...

event_type = common.ScEventType.ADD_OUTGOING_EDGE
event_params = ScEventParams(bounded_elem_addr, event_type, event_callback)
sc_event = client.events_create(event_params)
```

## is_event_valid

> Check is an event active or not.

Reviewer: ```suggestion
Check whether an event is active or not.
```

*Parameters*: An ScEvent class object.
*Returns*: The boolean value (true if an event is valid). 

```py
status = client.is_event_valid(sc_event)
status # True
```

## events_destroy
Destroy an event in the KB memory.

*Parameters*: ScEvent class objects.

> *Returns*: The boolean value (true if event was destroyed).

Reviewer: ```suggestion
*Returns*: The boolean value (true if an event was destroyed).
```

```py
status = client.events_destroy(sc_event)
status # true
```

# Classes
The library contains the python implementation of useful classes and functions to work with the sc-memory.

There is a list of classes:

 - ScKeynodes
 - ScAgent
 - ScModule

## ScKeynodes

> A singleton dictionary object which provides the ability to cache identifier and ScAddr of keynodes stored in the KB.

Reviewer: ```suggestion
A singleton dictionary object which provides the ability to cache the identifier and ScAddr of keynodes stored in the KB.
```

> Create an instance of ScKeynodes class to get access to the cache:

Reviewer: ```suggestion
Create an instance of the ScKeynodes class to get access to the cache:
```

```py
keynodes = ScKeynodes()
```

> Get provided identifier:

Reviewer: ```suggestion
Get the provided identifier:
```

```py
keynodes["identifier_of_keynode"] # returns an ScAddr of the given identifier
keynodes["not_stored_in_kb"] # returns an invalid ScAddr if an identifier does not exist in the memory
```

Use _resolve_identifiers()_ to upload identifiers from _Enum_ classes:

```py
class CommonIdentifiers(Enum):
    RREL_ONE = "rrel_1"
    RREL_TWO = "rrel_2"

class QuestionStatus(Enum):
    QUESTION_INITIATED = "question_initiated"
    QUESTION_FINISHED = "question_finished"
    QUESTION_FINISHED_SUCCESSFULLY = "question_finished_successfully"
    QUESTION_FINISHED_UNSUCCESSFULLY = "question_finished_unsuccessfully"
    
keynodes.resolve_identifiers([QuestionStatus, CommonIdentifiers])
```

## ScAgent
A class for handling a single ScEvent. Define your agents like this:
```py
class MyAgent(ScAgent):
    action = "Identifier_of_action_class"
    
    def register(self) -> ScEvent:

>         # override method, must return ScEvent instance

Reviewer: ```suggestion
        # override method, must return an ScEvent instance
```

        params = [
            MyAgent.keynodes["action_initiated"],
            common.ScEventType.ADD_OUTGOING_EDGE,
            MyAgent.run_impl
        ]
        event_params = ScEventParams(*params)
        sc_event = client.events_create(event_params)
        return sc_event[0]
    
    @staticmethod
    def run_impl(action_class: ScAddr, edge: ScAddr, action_node: ScAddr) -> None:
        # override method, must have 3 args and be static
        ...

```

## ScModule

> A class for handling multiple ScAgent. Define your modules like this:

Reviewer: ```suggestion
A class for handling a multiple ScAgent. Define your modules like this:
```
